src/Database/Redis/Cluster.hs

```diff
@@ -263,17 +263,28 @@ evaluatePipeline refreshShardmapAction conn@(Connection shardNodeVar infoMap _)
         -- otherwise if there is any response in the connection buffer it'll get forwarded to other requests that are reusing the same connection.
         -- leading to jumbled up responses
         resps <- concat <$>
-          mapM (\(resp, (cc, r)) -> case resp of
-              Right v -> return v
-              Left (err :: SomeException) ->
-                case fromException err of
-                  Just (er :: TimeoutException) -> refreshShardmapAction >> throwIO er
-                  _ -> getRandomConnection cc conn >>= (`executeRequests` r)
+          mapM (\(resp, (nc, r)) -> do
+                responses <-  case resp of
+                                Right v -> return v
+                                Left (err :: SomeException) ->
+                                    case fromException err of
+                                        Just (er :: TimeoutException) -> hasLocked $ refreshShardmapAction Nothing >> throwIO er
+                                        _ -> getRandomConnection nc conn >>= (`executeRequests` r)
+                refreshedShardMapAndNodeConnsIORef <- IOR.newIORef Nothing
+                mapM (\completedRequest@(CompletedRequest index request response) -> 
+                    case response of
+                        (Error errString) | B.isPrefixOf "MOVED" errString -> CompletedRequest index request <$> refreshShardMapAndRetryRequest refreshedShardMapAndNodeConnsIORef (hasLocked $ refreshShardmapAction (Just nc)) request
```

Comment:
if two redis calls are sent in one pipeline, wouldn't the refresh happens twice with this code?

Reply:
but calling of refresh function inside a `map` if it have multiple responses  with moved, wouldn't that invoke multiple refresh calls?

Reply:
even then one refresh is enough. because ideally one refresh will get all new nodes information. No need to run multiple times.

Reply:
another thing is on one refresh, the shardmap will be updated (should be updated) then you don't need any more refreshes in that batch.